src/cockpit/channels/pcp.py

```diff
@@ -0,0 +1,672 @@
+# * instances - the multi-instances to show (for example only `/dev/sda`)
+#
+# Metrics
+#
+# When opening the metrics channel you can specify the metrics you are
+# interested in, a PCP metric is described in pmapi as pmDesc:
+#
+# class pmDesc:
+# pmid  - unique ID of the metric
+# type  - data type (PM_TYPE_*)
+# indom - the instance domain
+# sem   - semantics of the value
+# units - dimension and units
+#
+# Important here are the type, Cockpit only supports PM_TYPE_DOUBLE, and integers as PM_TYPE_U64.
+#
+# The instance domain denotes if the metric is has multiple instances, for
+# example a disk metric can represent data for multiple disks.
+#
+# See `pminfo -f "kernel.all.load"` for example:
+# kernel.all.load
+#   inst [1 or "1 minute"] value 0.5
+#   inst [5 or "5 minute"] value 0.68000001
+#   inst [15 or "15 minute"] value 0.75999999
+#
+# These metrics are delivered to the JavaScript client as a list of values, the
+# meta message contains the list of names of instances for the UI to show.
+
+import asyncio
+import ctypes
+import glob
+import json
+import logging
+import platform
+import sys
+import time
+from typing import TYPE_CHECKING, Any, Dict, Iterable, List, NamedTuple, Optional, Sequence, Union
+
+from cockpit.protocol import CockpitProblem
+
+from ..channel import AsyncChannel, ChannelError
+from ..jsonutil import JsonObject, JsonValue, get_int, get_objv, get_str, get_strv
+
+if TYPE_CHECKING:  # pragma: no cover
+    import cpmapi as c_api
+    from pcp import pmapi
+else:
+    pmapi = None
+    c_api = None
+
+logger = logging.getLogger(__name__)
+
+
+class MetricNotFoundError(CockpitProblem):
+    pass
+
+
+class PcpMetricInfo(Dict[str, JsonValue]):
+    def __init__(self, value: JsonObject) -> None:
+        self.name = get_str(value, 'name')
+        self.derive = get_str(value, 'derive', '')
+        self.units = get_str(value, 'units', '')
+        super().__init__(name=self.name, derive=self.derive, units=self.units)
+
+
+class MetricInfo(NamedTuple):
+    pmid: int
+    derive: str
+    desc: Any
+    name: str
+    factor: float
+    units: Any
+    instanced: bool  # Multi instance metric
+
+
+def try_import_pcp() -> None:
+    global c_api, pmapi
+    if c_api is None or pmapi is None:
+        try:
+            import cpmapi as c_api
+            from pcp import pmapi
+        except ImportError as exc:  # pragma: no cover
+            raise ChannelError('not-supported', message='python3-pcp not installed') from exc
+
+
+class ArchiveInfo:
+    metric_descriptions: List[MetricInfo]
+
+    def __init__(self, context: 'pmapi.pmContext', start: float, path: str) -> None:
+        self.context = context
+        self.start = start
+        self.path = path
+        self.metric_descriptions = []
+
+    def sort_key(self) -> float:
+        return self.start
+
+
+class PcpMetricsChannel(AsyncChannel):
+    payload = 'metrics1'
+
+    pcp_dir: str
+    archive_batch: int = 60
```

Comment:
This is an instance variable, not a class one. Its initialization must go into `__init__()` or somewhere similar, like `parse_options()`. I absolutely despise this pattern of declaring instance variables as class ones, and I've had long arguments with Lis about it. It's just barely legit for initializing variables to `None`, but not with actual values please. (same for the timestamps).

It *should* be ok as parse_options re-initializes these, but doing this here is error prone and confusing. Please drop al initializers and just declare the types.

Do we have any test which opens more than one PCP channel at the same time?

Reply:
parse_options() is fine, thanks!